crates/typst-pdf/src/font.rs

```diff
@@ -23,6 +26,110 @@ const SYSTEM_INFO: SystemInfo = SystemInfo {
+                // create a fake page context for write_frame
+                // we are only interested in the contents of the page
+                let size = color_glyph.image.size();
+                let mut page_ctx = PageContext::new(ctx, page_ref, size);
+                page_ctx.bottom = size.y.to_f32();
+                page_ctx.content.start_color_glyph(width);
+                page_ctx.transform(Transform {
+                    sx: Ratio::one(),
+                    ky: Ratio::zero(),
+                    kx: Ratio::zero(),
+                    sy: size.aspect_ratio() * -1.0,
+                    tx: Abs::zero(),
+                    ty: size.y,
+                });
+                write_frame(&mut page_ctx, &color_glyph.image);
+                let stream = page_ctx.content.finish();
+                ctx.pdf.stream(page_ref, &stream);
+
+                glyphs_to_instructions.insert(cid, page_ref);
+            }
+
+            let mut pdf_font = ctx.pdf.type3_font(*subfont_id);
+            pdf_font.pair(Name(b"Resources"), ctx.type3_font_resources_ref);
+            pdf_font.bbox(font.bbox);
+            let scale_factor = font_info.ttf().units_per_em() as f32;
+            pdf_font.matrix([1.0 / scale_factor, 0.0, 0.0, 1.0 / scale_factor, 0.0, 0.0]);
+            let mut char_procs = pdf_font.char_procs();
+            for (gid, instructions_ref) in &glyphs_to_instructions {
+                char_procs
+                    .pair(Name(eco_format!("glyph{gid}").as_bytes()), *instructions_ref);
+            }
+            char_procs.finish();
+
+            let mut encoding = pdf_font.encoding_custom();
+            let mut differences = encoding.differences();
+            for gid in glyphs_to_instructions.keys() {
+                differences.consecutive(
+                    *gid as u8,
+                    vec![Name(eco_format!("glyph{gid}").as_bytes())],
+                );
+            }
+            differences.finish();
+            encoding.finish();
+
+            let first = 0;
+            let last = (end - 1 - start) as u8;
+            pdf_font.first_char(first);
+            pdf_font.last_char(last);
+            pdf_font.widths(widths);
+            pdf_font.to_unicode(cmap_ref);
+            pdf_font.font_descriptor(descriptor_ref);
+            pdf_font.finish();
+
+            // Encode a CMAP to make it possible to search or copy glyphs
+
+            // To avoid maintaining a separate glyph set structure for Type3
+            // fonts, a hack is used: the glyph ID is only in the
+            // less-significant byte (as there can only be 256 of them per
+            // font), the rest stores the Type3 font Ref.
+            let full_glyph_set = ctx.glyph_sets.get_mut(&font_info).unwrap();
+            // We retrieve only the part relevant for this font
+            let glyph_set = full_glyph_set
+                .iter()
+                .filter(|(k, _)| *k / 256 == subfont_id.get() as u16)
```

Comment:
Does this really work properly? What if the Type3 Refs exceed 256?

Reply:
I changed it so that it uses the index of the font in the list of all Type3 fonts. So the bug still exists, but only in case someone needs more than 256 Type3 fonts, which seems unlikely. If you are still unhappy with that, I can build something more solid.